### DISCUSSION TITLE
For generating a new column from two given columns, how can I reference these columns by index?

### DISCUSSION THREAD

**Discussion by halloleo** [saulpw/visidata#1110] [Q&A] [answered]
I have rather often sheets where a few headers are not unique, so in order to generate new columns (via the `=` command) I need to reference given columns via their index and not by name.

How can I do this?

**Comment by geekscrapy** [comment, marked as answer]
Hey!

One (longish) way to do this is with
`=visibleCols[1].getValue(row)`

This gets the *visible* column by index (in this case 1), and then gets the raw value of that row. There are other ways to get the data, based on what you want and speed (getTypedValue, getDisplayValue etc.), see here: https://www.visidata.org/docs/api/columns#visidata.Column.getValue

**Reply by saulpw** [reply]
Good idea, @geekscrapy !  I was just going to suggest renaming the columns :)

**Reply by halloleo** [reply]
Thanks for chiming! 

`=visibleCols[1].getValue(row)` feels pretty complicated. Also, when I hide some columns, the indexing is out of whack tothe index in the columns sheet. (I tend to add via `i` an index column to the Columns sheet.) 

So I guess I'll go for the column renaming. Then I can just say `=OneCol+" "+MyOtherCol`.

**Reply by saulpw** [reply]
Renaming is the best option, I think.  But you could use `columns[i].getValue(row)` instead, if you wanted the index to be in sync with the columns sheet.

**Reply by halloleo** [reply]
Cool! I'll keep that in mind. Speaking of an index on the Columns sheet: Is there a way to automatically add this index (so that not everytime I look at my columns list I have to press `i` to add the numbering)?

It's just a question. - After al in VisiData are so many things customisable... 😄

**Reply by saulpw** [reply]
You can't currently, but I've wondered about an option for a 'row number' column fixed at the far left of the window.  Then you could set the option on all ColumnsSheet.  Do you want it for readability?

**Reply by halloleo** [reply]
Well, yes and no: Yes, because in sheets with lots of columns it helps you to know better where which column is, but it also helps a lot for interoperability with other tools: 

When I need to do some processing in external tools (which expect column indexes) I have started to use `vd` as a "column viewer". (In the past I used `csvcut -n` for this, but `vd` in a new xterm seems to be more practical.)

I hope this makes sense.

**Reply by saulpw** [reply]
It does, but I'm not sure what you do with that column number.  Do you type it manually into one of those tools in another window?  Or is there some way you need the column number to be more than cosmetic (like usable in a VisiData expression)?

**Reply by halloleo** [reply]
My _main_ use is indeed only to use it somewhere else by typing or to check columns are in the right place.

But I think it can come in handy _inside_ VisiData as well: Sometimes you don't care about column names, but more about column indexes: Say you have column sequences like `Street, City, State, AltStreet, AltCity, AltState, ClientStreet, ClientCity, ClientState`, then you know `State` is always `Street` index + 2.
